### DISCUSSION TITLE
How to use GeneralContact to simulate an arbitrary floor in contact with wheels

### DISCUSSION THREAD

**Discussion by ManuelZ** [jgerstmayr/EXUDYN#65] [Q&A on usage of Exudyn] [answered]
I have created this simulation:

https://github.com/jgerstmayr/EXUDYN/assets/115771/2a482273-22af-479d-90da-a9ef504581e9

Using `ObjectConnectorRollingDiscPenalty` between the wheels and the floor:
```
self.data = mbs.AddNode(
    NodeGenericData(initialCoordinates=[0, 0, 0], numberOfDataCoordinates=3)
)

stiffnes = 1e9
damping = stiffnes * 0.01

self.rolling = mbs.AddObject(
    ObjectConnectorRollingDiscPenalty(
        markerNumbers=[marker_ground, self.rigid_body_markers["center"]],
        nodeNumber=self.data,
        discRadius=self.r,
        discAxis=[0, 0, 1],
        planeNormal=[0, 1, 0],
        dryFriction=[0.4, 0.4],
        dryFrictionProportionalZone=1e-1,
        rollingFrictionViscous=0.2 * 0,
        contactStiffness=stiffnes,
        contactDamping=damping,
        visualization=rolling_disk_visualization,
    )
)
```

But now I would like to use an arbytrary floor, made out with custom shapes. To do so, I believe I need to use the `GeneralContact` module, but I'm not being able to make it work:

https://github.com/jgerstmayr/EXUDYN/assets/115771/47017ff8-f1a5-4d5e-a1d7-ac296e10691f

```

contact_stiffness = 5000
contact_damping = 1e-4 * contact_stiffness

gContact = mbs.AddGeneralContact()

[meshPoints, meshTrigs] = GraphicsData2PointsAndTrigs(ground.graphics)
[meshPoints, meshTrigs] = RefineMesh(meshPoints, meshTrigs)

gContact.AddTrianglesRigidBodyBased(
    rigidBodyMarkerIndex=ground.rigid_body_markers["center"],
    contactStiffness=contact_stiffness,
    contactDamping=contact_damping,
    frictionMaterialIndex=0,
    pointList=meshPoints,
    triangleList=meshTrigs,
)

gContact.AddSphereWithMarker(
    wheel_left.rigid_body_markers["center"],
    radius=wheel_left.r,
    contactStiffness=contact_stiffness,
    contactDamping=contact_damping,
    frictionMaterialIndex=0,
)

gContact.AddSphereWithMarker(
    wheel_right.rigid_body_markers["center"],
    radius=wheel_right.r,
    contactStiffness=contact_stiffness,
    contactDamping=contact_damping,
    frictionMaterialIndex=0,
)

# search tree size
ssx = 3
ssy = 3
ssz = 3

mu = 0.5
gContact.verboseMode = 1
gContact.frictionProportionalZone = 1e-5
gContact.SetFrictionPairings(mu * np.eye(1))
gContact.SetSearchTreeCellSize(numberOfCells=[ssx, ssy, ssz])

```

The chassis has a mass of 1500 kg, If I increase the `contactStiffness`, the wheels don't penetrate the floor that much, but still do. If I keep increasing the stiffness, I get more and more `Newton (time/load step #11115): convergence failed after 25 iterations;` errors.

What parameters should I tune? Am I doing something wrong?

In both cases I'm using `exu.DynamicSolverType.GeneralizedAlpha` because I'm using joints.

**Comment by jgerstmayr** [comment]
GeneralContact is not capable of using implicit solvers (GeneralizedAlpha, etc.) for contact of spheres with triangles. This is on the todo list. It only works with explicit solvers, thus being restricted to moderately stiff systems.
You can still use joints, if you apply KinematicTree objects for the joints.
if you intend to have a 2D ground floor, you can use (fixed/constrained) ANCF cables for the ground. This also works with implicit solvers.
However, the GeneralContact will be probably available in June 2024 for the implicit solver as well (only one C++ function and some tests missing...)

**Comment by ManuelZ** [comment]
Thank you.

Using the latest Exudyn code, I have been able to make the simulation work in certain cases with GeneralContact with an implicit solver (GeneralizedAlpha), with an arbitrary STL floor. However I'm having trouble when adding obstacles.

In this simple example, the simulation keeps working after the disk makes contact with the obstacle , but with a lot of these : 
```
Newton (time/load step #38381): convergence failed after 25 iterations; relative error = 0.000815751, time = 3.5785
```

(The reddish lines in the video come from the contact visualization settings)

https://github.com/jgerstmayr/EXUDYN/assets/115771/020adea8-1483-4d94-b2b0-b37eb68557e8


But in a more complex simulation, it doesn't get past that point:

https://github.com/jgerstmayr/EXUDYN/assets/115771/b7fa1e6c-8533-4f4b-9490-576972978bf8

At that point the simulation crashes. With less pronounced discontinuities in the floor, like a sudden small ramp with an angle of 2 deg, I can get past that point increasing  `simulationSettings.timeIntegration.newton.maxIterations` from the default 25 to 50 or 100.

For the last simulation I'm using these parameters:
```
ContactConfiguration(
    friction=0.3,
    search_tree=ContactSeachTreeCellSizes(ssx=100, ssy=5, ssz=1),
    friction_proportional_zone=1e-5,
    contact_stiffness=1e9,  # for AddTrianglesRigidBodyBased
    contact_damping=1e5,  # for AddTrianglesRigidBodyBased
)
```

But at this point I'm just turning knobs without really knowing what to turn.

Is there something I could improve?

**Comment by jgerstmayr** [comment]
The implicit solver is still not fully working. Thus, only the explicit solver will work, requiring that there are no constraints in the system.

**Comment by ManuelZ** [comment]
Would the option of using ANCF cables for the ground with an implicit solver, constraints and `RollingDisc`/`RollingDiscPenalty` work? Or would I face the same problem that occurs when using `GeneralContact` plus an implicit solver?

**Reply by jgerstmayr** [reply, marked as answer]
In a 2D-case, you can use ACNF cables fixed to ground, with spheres rolling on them. This works with the implicit solver in this case.

**Comment by ManuelZ** [comment]
Using ANCF cables with `GeneralContact` and an implicit solver worked perfectly! Thank you!

https://github.com/jgerstmayr/EXUDYN/assets/115771/0657a9ab-9fdd-4dfa-9152-2f814f1adbe2
